kubernetes-simulator/simulator
# Scoring

This implements the functionality decribed in #173 ... the attack container has been pushed with the tag `scoring-mvp`.

**To test**

* Clone and checkout the scoring branch
* Recreate your infra with `--attack-container-tag=scoring-mvp`
* Check each of the logic branches described in the ticket #173 

Try starting tasks showing hints starting new tasks, going back and forth etc and make sure the messaging feels right and there are no errors etc :)

Other than obvious bugs, Im particularly interested in feedback on the user experience and messaging.  Unleash your creative destruction :)

This resolves both #173  and partially #172 (this implementation does not ask the user to confirm they want to incur the penalty to see the hint)

### Update

I have done a large refactoring to make the code clearer and to add a boatload of automated tests to check all the edge cases I could think of.  Unfortunately this makes this a bit of a monster pull request, but the behaviour is the same and the cases I have covered can be seen by running the tests in verbose mode `cd tools/scenario-tools/ && npm test -- --verbose`:

```
<snip>
  ✔ tasks_test › getCurrentTask returns the current task
  ✔ tasks_test › getCurrentTask returns undefined when no current task
  ✔ tasks_test › updateProgressWithNewTask sets current_task
  ✔ tasks_test › updateProgressWithNewTask initialises task progress
  ✔ tasks_test › updateProgressWithNewTask does not overwrite existing progress
  ✔ tasks_test › processResponse throws when answer is not defined
  ✔ tasks_test › processResponse returns false when answer is cancel
  ✔ tasks_test › processResponse sets score to skip when answer is no
  ✔ tasks_test › processResponse sets score when answer is yes
  ✔ tasks_test › processTask warns for invalid task and returns false
  ✔ tasks_test › endTask warns if the user has not started a task
  ✔ tasks_test › startTask warns if user tries to start the same task
  ✔ tasks_test › startTask initialises progress when user has no progress
  ✔ tasks_test › startTask updates current_task and doesnt rescore if already scored
  ✔ tasks_test › processTask writes current_task to progress
  ✔ tasks_test › endTask returns false if the user cancels
  ✔ tasks_test › endTask skips scoring and ends task when user says no to scoring
  ✔ tasks_test › endTask sets score and ends task when user says yes to scoring
  ✔ tasks_test › startTask updates current_task and skips scoring if answer is no
  ✔ tasks_test › startTask updates current_task and scores if answer is yes
<snip>
```
